### DISCUSSION TITLE
Can a Graph Index be used in the middle of traversal?

### DISCUSSION THREAD

**Discussion by roi-tevel** [JanusGraph/janusgraph#3236] [Q&A]
Hi, 
Assuming we have such traversal: 
`g.V().has('first_indexed_prop', 'first_value').out('some_edge').out('some_other_edge').has('other_indexed_prop','other_value').out().count().next()`
will JanusGraph be able to use the index we have on `other_indexed_prop` of the Vertex? (when I run it with profile() it seems as if it doesn't use it, even when its a mixed index)
or do we need to add other_indexed_prop as a property of the edge (some_other_edge) and add an edge index in order for JG to be able to use another index in the middle of a traversal? 
Thanks, 
Roi

**Comment by vtslab** [comment]
This is discussed in [https://docs.janusgraph.org/schema/index-management/index-performance/#using-vertex-centric-indexes-on-adjacent-vertex-ids](https://docs.janusgraph.org/schema/index-management/index-performance/#using-vertex-centric-indexes-on-adjacent-vertex-ids)

Retrieving vertices further in a traversal before filtering is not very expensive (compared to performing a full table scan), unless your graph has lots of "other edges". In the latter case, the link above provides a solution, at the cost of some additional code.

**Reply by roi-tevel** [reply]
Thanks, our issue is that we have something like
Company -> Job -> Execution 
each company has lots of jobs, each job has lots of executions, but only one marked as 'latest' ('latest' being currently a property of the Execution Vertex).
We want to traverse from company to all its latest job executions. and the filtering is quite expensive (takes around 13 seconds), and it seems like we can not use the solution discussed in this link, since there are lots of 'latest' executing jobs in different companies. so if I understand correctly our best options are either:
1. to make 'latest' a property of the edge between Job and Execution, or
2. to have 'company_id' in addition to latest as properties of Execution, and then start the traversal from Execution? (where we can have an index on 'company_id' + 'latest')

Thanks again, 
Roi

**Reply by li-boxuan** [reply]
If there is only one execution marked as "latest", can you make the edge between job and latest execution a separate edge label?

**Reply by roi-tevel** [reply]
Actually our problem is a bit more complicated, we have also Container Vertices which are linked to Execution, through multiple edge labels and sometimes we need to traverse from Container to the latest Execution (filtering out containers of previous executions) , so if we go with separate edge labels we will need to add multiple edge_label_latest edges from Execution to the different Container types + one for the Job. 

btw, I noticed that when putting 'latest' inside of a where clause (instead of has step)` .where(__.values('latest').is(eq(true)) `performance improved dramatically compared to `.has('latest', true)`. from 13 seconds, to around 600 milliseconds. not sure why though?

**Reply by li-boxuan** [reply]
Hey, if you add `profile()` step at the end of your query, it will show you the actual steps JanusGraph takes to execute the query. That might answer your question.

**Reply by roi-tevel** [reply]
Hey, thanks, from the profile it seems that it is doing  TraversalFilterStep in case of where, and that this is somehow much much more efficient than the HasStep step, but I am not sure why?

this is with has:
```
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[search_name.eq(amazon)])                            1           1          86.283     0.25
  constructGraphCentricQuery                                                                   0.048
  GraphCentricQuery                                                                           86.098
    \_condition=(search_name = amazon)
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=[(search_name = amazon)]:bySearchName
    \_index=bySearchName
    \_index_impl=search
    backend-query                                                      1                      86.026
    \_query=bySearchName:[(search_name = amazon)]:bySearchName
JanusGraphMultiQueryStep                                               1           1           0.039     0.00
JanusGraphVertexStep(OUT,[subject],vertex)                             7           7          10.495     0.03
    \_condition=type[subject]
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=subject:SliceQuery[0x701AC0,0x701AC1)
    \_multi=true
    \_vertices=1
  optimization                                                                                 0.058
  backend-query                                                        7                      10.293
    \_query=subject:SliceQuery[0x701AC0,0x701AC1)
JanusGraphMultiQueryStep                                               7           7           0.047     0.00
JanusGraphVertexStep(OUT,[processed_by],vertex)                     8180        8180          44.067     0.13
    \_condition=type[processed_by]
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=processed_by:SliceQuery[0x701DE0,0x701DE1)
    \_multi=true
    \_vertices=7
  optimization                                                                                 0.051
  backend-query                                                     8180                      30.742
    \_query=processed_by:SliceQuery[0x701DE0,0x701DE1)
NoOpBarrierStep(2500)                                               8180        8180          22.035     0.06
HasStep([latest.eq(true)])                                                                 33942.107    99.52
CountGlobalStep                                                        1           1           0.041     0.00
                                            >TOTAL                     -           -       34105.117        -
```
this is without has:
```
==>Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[search_name.eq(amazon)])                            1           1           0.232     0.48
  constructGraphCentricQuery                                                                   0.069
  GraphCentricQuery                                                                            0.097
    \_condition=(search_name = amazon)
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=[(search_name = amazon)]:bySearchName
    \_index=bySearchName
    \_index_impl=search
JanusGraphMultiQueryStep                                               1           1           0.031     0.06
JanusGraphVertexStep(OUT,[subject],vertex)                             7           7           0.102     0.21
    \_condition=type[subject]
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=subject:SliceQuery[0x701AC0,0x701AC1)
    \_multi=true
    \_vertices=1
  optimization                                                                                 0.026
JanusGraphMultiQueryStep                                               7           7           0.024     0.05
JanusGraphVertexStep(OUT,[processed_by],vertex)                     8180        8180           4.701     9.63
    \_condition=type[processed_by]
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=processed_by:SliceQuery[0x701DE0,0x701DE1)
    \_multi=true
    \_vertices=7
  optimization                                                                                 0.005
JanusGraphMultiQueryStep                                            8180        8180           6.187    12.67
NoOpBarrierStep(2500)                                               8180        8180           4.803     9.84
TraversalFilterStep([JanusGraphPropertiesStep([...                    55          55          32.597    66.77
  JanusGraphPropertiesStep([latest],value)                            55          55          22.991
    \_condition=type[latest]
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=latest:SliceQuery[0x5A80,0x5A81)
    \_multi=true
    \_vertices=8180
    optimization                                                                               0.024
  IsStep(eq(true))                                                                             3.724
CountGlobalStep                                                        1           1           0.142     0.29
                                            >TOTAL                     -           -          48.823        -
```

**Reply by roi-tevel** [reply]
> This is discussed in https://docs.janusgraph.org/schema/index-management/index-performance/#using-vertex-centric-indexes-on-adjacent-vertex-ids
> 
> Retrieving vertices further in a traversal before filtering is not very expensive (compared to performing a full table scan), unless your graph has lots of "other edges". In the latter case, the link above provides a solution, at the cost of some additional code.

Thanks again, this also applies for using an index for ordering? Janus can not use the index for ordering if the order().by() is further in the traversal? it there a way around that?

**Reply by li-boxuan** [reply]
@roi-tevel The performance improvement you saw is due to multithreading. In the second profile, you can see ` \_multi=true` in `JanusGraphPropertiesStep`, indicating that it uses multi-threading. Theoretically, we could (and we should) leverage multi-threading for the `HasStep` in the first query as well, but that is something not done yet (created https://github.com/JanusGraph/janusgraph/issues/3244). Btw what JanusGraph version you are using?

**Reply by roi-tevel** [reply]
@li-boxuan Thanks, good to know. we are using version 0.6.0

**Reply by vtslab** [reply]
> Janus can not use the index for ordering if the order().by() is further in the traversal? it there a way around that?

See: https://docs.janusgraph.org/v0.4/index-management/index-performance/#ordering
For mixed indices the order().by() modifier is pushed down to the indexing backend. This should work for successive has().order().by() steps, I am not sure what happens if the order().by() does not immediately follow the has() step.
